### PR TITLE
Add new cancel-generation and switch-generation cmd. Apiserver implementation to follow.

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

Initial implementations of the switch-generation and cancel-generation commands. Apiserver side of the feature is not implemented.

## QA steps

1. export JUJU_DEV_FEATURE_FLAGS=generations
2. juju cancel-generation
changes dropped and target generation set to current <-- result
4. juju help cancel-generation
4. juju switch-generation next
changes dropped and target generation set to next <-- result
4. juju switch-generation current 
changes dropped and target generation set to current <-- result

### CODE REVIEW

File: cmd/juju/model/addgeneration.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ Examples:
     juju add-generation
 
 See also:
-	cancel-generation
+    cancel-generation
```
Comment (hmlanigan) [inline cmd/juju/model/addgeneration.go]: drive by

File: cmd/juju/model/cancelgeneration.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+	"github.com/juju/juju/api/modelgeneration"
+	jujucmd "github.com/juju/juju/cmd"
+	"github.com/juju/juju/cmd/modelcmd"
+)
+
+const (
+	cancelGenerationSummary = "Cancels the new generation to the model."
+	cancelGenerationDoc     = `
+Cancel the next generation. This will abort anything in the next
```
Comment (hmlanigan) [inline cmd/juju/model/cancelgeneration.go]: there is a card to refine the help text here: https://trello.com/c/VBlUfGju/925-refine-text-for-cancel-generation-cmd-help

File: cmd/juju/model/switchgeneration.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+	"github.com/juju/juju/api/modelgeneration"
+	jujucmd "github.com/juju/juju/cmd"
+	"github.com/juju/juju/cmd/modelcmd"
+)
+
+const (
+	switchGenerationSummary = "Switch to the given generation."
+	switchGenerationDoc     = `
+Switch to the given generation, either current or next.
```
Comment (hmlanigan) [inline cmd/juju/model/switchgeneration.go]: there is a card to refine the help text here: https://trello.com/c/Kff2V2LS/924-refine-text-for-switch-generation-cmd-help

**Comment by hmlanigan** [comment]
There is a lot of copy paste in this work.  It appears I made all the changes necessary...

File: cmd/juju/model/switchgeneration_test.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+func (s *switchGenerationSuite) runCommand(c *gc.C, api model.SwitchGenerationCommandAPI, args ...string) (*cmd.Context, error) {
+	cmd := model.NewSwitchGenerationCommandForTest(api, s.store)
+	return cmdtesting.RunCommand(c, cmd, args...)
+}
+
+func setUpSwitchMocks(c *gc.C) (*gomock.Controller, *MockSwitchGenerationCommandAPI) {
+	mockController := gomock.NewController(c)
+	mockSwitchGenerationCommandAPI := NewMockSwitchGenerationCommandAPI(mockController)
+	mockSwitchGenerationCommandAPI.EXPECT().Close().Times(1)
```
Comment (manadart) [inline cmd/juju/model/switchgeneration_test.go]: `.Times(1)` implicit here.

**Comment by manadart** [comment]
!!build!!

File: cmd/juju/model/switchgeneration_test.go (hunk 4 above)
Comment (hmlanigan) [reply cmd/juju/model/switchgeneration_test.go]: updated

**Comment by hmlanigan** [comment]
failed with: WARNING juju.testing mgo.go:212 failed to start mongo: mongod failed to listen on port 37828
$$merge$$

**Pull request merged**